src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs

```diff
@@ -968,21 +968,40 @@ private unsafe bool TryGetOrCreateObjectForComInstanceInternal(
                             _rcwCache.Remove(identity);
                         }
                     }
+                }
+
+                if (wrapperMaybe is not null)
+                {
+                    retValue = wrapperMaybe;
 
-                    if (wrapperMaybe is not null)
+                    using (_lock.EnterScope())
                     {
-                        retValue = wrapperMaybe;
-                        NativeObjectWrapper wrapper = NativeObjectWrapper.Create(
-                            identity,
-                            inner,
-                            this,
-                            retValue,
-                            flags);
-                        if (!s_rcwTable.TryAdd(retValue, wrapper))
+                        if (!s_rcwTable.TryGetValue(retValue, out NativeObjectWrapper? wrapper))
```

Comment:
Is the lock around `s_rcwTable.TryGetValue` necessary - can we enter the lock later for `_rcwCache.Add` only?